src/shogun/machine/GeneralizedLinearMachine.cpp

```diff
@@ -0,0 +1,299 @@
+			bias = bias - m_learning_rate * gradient_bias;
+	
+		//Apply proximal operator
+		w = apply_proximal_operator(w, m_lambda * m_alpha);
+
+		//Convergence by relative parameter change tolerance
+		float64_t norm_update = linalg::norm(linalg::add(w, w_old, 1.0, -1.0));
+		if(i > 0 && (norm_update / linalg::norm(w)) < m_tolerance)
+			break;		
+	}
+
+	//Update the estimated variables
+	m_w = w;
+	m_bias = bias;
+
+	return true;
+}
+
+//TODO
+std::shared_ptr<RegressionLabels> GeneralizedLinearMachine::apply_regression(std::shared_ptr<Features> data)
+{
+	
+}
+
+//TODO
+SGVector<float64_t> GeneralizedLinearMachine::apply_get_outputs(std::shared_ptr<Features> data)
+{
+
+}
+
+//TODO
+bool GeneralizedLinearMachine::train_machine(std::shared_ptr<const DenseFeatures<float64_t>> feats)
+{
+
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::get_w() const
+{
+	return m_w;
+}
+
+void GeneralizedLinearMachine::set_w(const SGVector<float64_t> w)
+{
+	m_w = w;
+}
+
+void GeneralizedLinearMachine::set_bias(float64_t b)
+{
+	m_bias=b;
+}
+
+float64_t GeneralizedLinearMachine::get_bias() const
+{
+ 	return m_bias;
+}
+
+void GeneralizedLinearMachine::set_features(std::shared_ptr<DotFeatures> feat)
+{
+ 	features=std::move(feat);
+}
+
+std::shared_ptr<DotFeatures> GeneralizedLinearMachine::get_features()
+{
+	return features;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::conditional_intensity(SGMatrix<float64_t> X, SGVector<float64_t> w, float64_t bias)
+{
+	SGVector<float64_t> z = compute_z(X, w, bias);
+	SGVector<float64_t> result = non_linearity(z);
+	return result;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::compute_z(SGMatrix<float64_t> X, SGVector<float64_t> w, float64_t bias)
+{
+	SGVector<float64_t> z = linalg::matrix_prod(X, w, false);
+	return z;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::non_linearity(SGVector<float64_t> z)
+{
+	SGVector<float64_t> result;
+	switch (distribution)
+	{
+	case POISSON:
+		result = SGVector<float64_t>(z);
+		float64_t bias = 0;
+
+		if(m_fit_intercept)
+			bias = (1 - m_eta) * exp(m_eta);
+
+		for (int i = 0; i < z.vlen; i++)
+		{
+			if(z[i]>m_eta)
+				result[i] = z[i] * exp(m_eta) + bias;
+			else
+				result[i] = exp(z[i]);
+		}
+		break;
+	
+	default:
+		break;
+	}
+	return result;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::gradient_non_linearity(SGVector<float64_t> z)
+{
+	SGVector<float64_t> result;
+	switch (distribution)
+	{
+	case POISSON:
+		result = SGVector<float64_t>(z);
+		for (int i = 0; i < z.vlen; i++)
+		{
+			if(z[i]>m_eta)
+				
+				result[i] = exp(m_eta);
+			else
+				result[i] = exp(z[i]);
+		}
+		break;
+	
+	default:
+		break;
+	}
+	return result;
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::compute_grad_L2_loss_w(SGMatrix<float64_t> X, SGVector<float64_t> y, SGVector<float64_t> w, float64_t bias)
+{
+	int32_t n_samples = y.vlen;
+	int32_t n_features = X.num_rows;
+	if(m_tau == NULL)
+		m_tau = SGMatrix<float64_t>::create_identity_matrix(w.vlen, 1.0);
```

Comment:
numerically it rately makes sense to create an identity matrix, instead use functions that add stuff to the diagonal of other matrices

Reply:
I did that so that I could then successfully make the InvCov matrix by multiplying it by its transpose whether or not the m_tau matrix is used.
What else can I do?

Reply:
I don't understand what you mean. But numerically, you don't do things like multiplying by an identity matrix....just extend the if-then-else and fill the inv covariance matrix in there. But then you again store an identity matrix. It is better to write code that branches for those cases